### PR TITLE
Add test harness in broccoli-middleware repo

### PR DESCRIPTION
**Pull request by kratiahuja**
This PR primarily addresses this [issue](https://github.com/ember-cli/broccoli-middleware/issues/7)

Adds some basic tests:
- [x] Basic file serving test
- [x] Error when file not found
- [x] Error when watcher is rejected
- [x] Add test for autoIndex flag
- [x] Added test around serving directory template

cc: @stefanpenner

### CODE REVIEW

**Comment by kratiahuja** [comment]
aah test failing because of `class`. Will switch to old school way. 

Should I also add code coverage?

**Comment by kratiahuja** [comment]
Going to add some more tests.

**Comment by kratiahuja** [comment]
Added enough tests for all basic functionality of this middleware.

**Comment by kratiahuja** [comment]
@stefanpenner This should be ready to be merged. I will add code coverage tool in another PR to isolate the changes and make it easy for review.

File: test/middleware-test.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+      server.start()
+        .then(function(info) {
+          return server.request('/index.html', {
+            info: info
+          });
+        })
+        .then(function(content) {
+          expect(content).to.match(/This is broccoli middleware page/);
+          done();
```
Comment (stefanpenner) [inline test/middleware-test.js]: rather then using `done`, which if an assertion before it fails (like line 42) will cause the test to timeout. Lets use mocha's build in promise support.

Simply return a promise in the `it` `function`, like:

```js
it('foo', function() {
  return server.start().then(...);
});
```

If the promise rejects, the test fails, if the promise fulfills the test passes.

File: test/middleware-test.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    it('responds with directory structure template if request is a directory and autoIndex is set to true', function(done) {
+      watcher['builder']['outputPath'] = fixture('no-index');
+      var middleware = broccoliMiddleware(watcher, {
+        autoIndex: true
+      });
+
+      server = new TestHTTPServer(middleware);
+
+      server.start()
```
Comment (stefanpenner) [inline test/middleware-test.js]: return the promise, rather then using `done`

File: test/middleware-test.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        .then(function(info) {
+          return server.request('', {
+            info: info
+          });
+        })
+        .then(function(content) {
+          expect(content).to.match(/This is broccoli middleware page/);
+          done();
+        })
```
Comment (stefanpenner) [inline test/middleware-test.js]: return the promise rather then using done

File: test/middleware-test.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      server.start()
+        .then(function(info) {
+          return server.request('/index.html', {
+            info: info
+          });
+        })
+        .then(function() {
+        }, function(error) {
+          expect(error).to.match(/StatusCodeError: 500/);
+          done();
```
Comment (stefanpenner) [inline test/middleware-test.js]: return the promise rather then using done

File: test/middleware-test.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+      server = new TestHTTPServer(middleware);
+
+      server.start()
+        .then(function(info) {
+          return server.request('/index.html', {
+            info: info
+          });
+        })
+        .then(function() {
```
Comment (stefanpenner) [inline test/middleware-test.js]: `.catch`

File: test/middleware-test.js [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        })
+    });
+  });
+
+  describe('watcher is rejected', function() {
+    var watcher;
+    var server;
+
+    beforeEach(function() {
+      watcher = new RSVP.Promise(function(resolve, reject) {
```
Comment (stefanpenner) [inline test/middleware-test.js]: A simpler way to describe this is:

```js
RSVP.Promise.reject({ stack: '..' , broccoliPayload: '..' })
```

**Review by stefanpenner — CHANGES_REQUESTED** [review]
Largely looks good, left some comments. Once addressed this should be mergable.

**Comment by stefanpenner** [comment]
It looks like travis is enabled, do you mind adding the badge to the readme? I believe it is:

```md
[![Build Status](https://travis-ci.org/ember-cli/broccoli-middleware.svg?branch=master)](https://travis-ci.org/ember-cli/broccoli-middleware)
```

**Comment by stefanpenner** [comment]
appveyor badge is:

```
[![Build status](https://ci.appveyor.com/api/projects/status/p4kvoah8x88dvwjo/branch/master?svg=true)](https://ci.appveyor.com/project/embercli/broccoli-middleware/branch/master)
```

also just enabled appveyor 👍

**Comment by kratiahuja** [comment]
@stefanpenner Thanks for the feedback. I totally forgot about the promise API of mocha 😄 . I have fixed it now.

One thing I noticed is the image for appveyor seems to be broken. Not sure if the link is correct or not.

File: test/middleware-test.js (hunk 4 above)
Comment (kratiahuja) [reply test/middleware-test.js]: Fixed.

File: test/middleware-test.js (hunk 5 above)
Comment (kratiahuja) [reply test/middleware-test.js]: Fixed.

File: test/middleware-test.js (hunk 6 above)
Comment (kratiahuja) [reply test/middleware-test.js]: Fixed

**Review by stefanpenner — APPROVED** [review]

**Comment by stefanpenner** [comment]
> One thing I noticed is the image for appveyor seems to be broken. Not sure if the link is correct or not.

Ah, ill fix it.

This looks great!

**Comment by stefanpenner** [comment]
@kratiahuja the badge seems to work, it may have been an appveyor glitch yesterday. They had some issue.

**Comment by kratiahuja** [comment]
Aah ok didn't realize something was broken there.

File: test/middleware-test.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+          return server.request('/index.html', {
+            info: info
+          });
+        })
+        .then(function(content) {
+          expect(content).to.match(/This is broccoli middleware page/);
+        });
+    });
+
+    it('responds with error if file not found', function(done) {
```
Comment (Turbo87) [inline test/middleware-test.js:44]: @kratiahuja unfortunately this doesn't really work. A test function must not use the `done` callback and return a `Promise` at the same time (see #13).

I tried changing it to the following, but it didn't seem to resolve properly:

```js
    it('responds with error if file not found', function() {
      var deferred = RSVP.defer();

      watcher['builder']['outputPath'] = fixture('basic-file');
      var middleware = broccoliMiddleware(watcher, {
        autoIndex: false
      });

      var wrapperMiddleware = function(req, resp /*next*/) {
        return middleware(req, resp, function(err) {
          expect(resp.finished).to.be.false;
          expect(err.message).to.have.string('ENOENT');
          deferred.resolve();
        })
      };
      server = new TestHTTPServer(wrapperMiddleware);

      return server.start().then(function(info) {
        return server.request('/non-existent-file', {
          info: info
        });
      }).then(function() {
        return deferred.promise;
      });
    });
```

Comment (kratiahuja) [reply test/middleware-test.js:44]: @Turbo87 : That's interesting. I am surprised how CI didn't catch it earlier unless mocha added that assertion in a later version. 

I'll debug and see if I can fix it.

Comment (kratiahuja) [reply test/middleware-test.js:44]: @Turbo87 : Attempt to fix it in a straightforward way : #14

Comment (Turbo87) [reply test/middleware-test.js:44]: Yes, they introduced that assertion in 3.0

**Pull request merged**